decorators.py: Improve yield_once speed

justuswilhelm opened the pull request:
Changes:
- yield_once uses a set instead of a list to lookup already seen items
- change the docstring to clarify that it does not wrap iterators but methods that return iterators
- a doctest verifies that it works.

In order to run the doctests, type

```
python -m doctest coala_decorators/decorators.py
```

I wasn't sure whether there is a CI unit test suite yet, so I just created a simple doctest. This might be related to issue #9. Please let me know if anything can be improved.

Makman2 commented:
As you have three changes, I think you can make three commits out of them :)
nice btw that you figured out a manual way for doing doctests until we set up our CI^^

justuswilhelm commented:
OK, I see how changing the docstring warrants its own commit, but surely, adding a doctest still belongs to this commit, right?

AbdealiJK commented:
I think what @Makman2 meant by the 3 changes was:
 - Simplify if-else by removing redundant `pass`
 - Use `set` instead of list for better performance
 - Improve documentation of yield_once (doctests are also a method to improve documentation :))

These are all separate and need not be in the same commit

AbdealiLoKo commented on coala_decorators/decorators.py:
```diff
@@ -6,16 +6,21 @@ def yield_once(iterator):
+    :return:         A method that returns an iterator
+                     that yields every result only once at most.
+
+    >>> @yield_once
```
We normally add doctests on top of the :param: stuff

justuswilhelm commented:
Ah! Gotcha.

Makman2 commented:
actually I meant
1. doctest
2. documentation text changes
3. code changes

but @AbdealiJK 's version is also okay, this makes sense too :+1:

sils1297 commented:
@justuswilhelm small changes needed, removing pending review label

justuswilhelm committed:
decorators.py: Add yield_once doctest

justuswilhelm committed:
decorators.py: Fix documentation

justuswilhelm committed:
decorators.py: Use set in yield_once

This will make `item in yielded` lookup perform in O(1)

justuswilhelm committed:
decorators.py: Remove else branch in yield_once

justuswilhelm committed:
decorators.py: improve yield_once wrapped method

Using the `wraps` decorator in functools allows us to preserve the
wrapped's name and docstring.

justuswilhelm commented:
I hope I've captured all your requirements :) Thanks for waiting, I had a long week.
As always, run
```
python -m doctest coala_decorators/decorators.py
```
to perform tests.

sils1297 commented:
@Adrianzatreanu can you review?

sils1297 commented:
Do we want to move this repo to GitLab and see how CI setup is going on GitLab? I think it'd make things easier

justuswilhelm commented:
Sounds like a good idea. If it turns out to be a lot easier, would it make sense to start moving other repos over to GitLab as well?

sils1297 commented:
ack https://github.com/coala-analyzer/coala-decorators/pull/15/commits/0ecb35eb9ff276aeb127cff5a6d39c6970f52171 https://github.com/coala-analyzer/coala-decorators/pull/15/commits/19b6ecaf07756eb22de6d22d3795ba80b992f00e https://github.com/coala-analyzer/coala-decorators/pull/15/commits/eeaee97fefb990841c3b18c5a1e6f8232160a115 https://github.com/coala-analyzer/coala-decorators/pull/15/commits/b53e93f660c8cb752eed2bd27553e87de0598ac7 https://github.com/coala-analyzer/coala-decorators/pull/15/commits/e914ec5de7193fb01067f25dc0f4ba6e2ffec3cb

justuswilhelm commented:
@sils1297 Thanks!

sils1297 commented:
@justuswilhelm nice work :)

sils1297 commented:
@justuswilhelm rebase?

justuswilhelm commented:
Rebase? Is the branch not up to date anymore?

sils1297 commented:
it's not, adrian merged a commit adding the VERSION file to package_data

justuswilhelm commented:
@sils1297 done. I wish I was able to see that information on GH as well. Only thing I get to see is that
> You're not authorized to merge this pull request

sils1297 commented:
@justuswilhelm now do a `git log --oneline master..` and paste it together with a `reack` into the github comment. reack must be at start or end of the comment

justuswilhelm commented:
reack d81754b decorators.py: improve yield_once wrapped method
12ffb0b decorators.py: Remove else branch in yield_once
ba356c5 decorators.py: Use set in yield_once
2e23c40 decorators.py: Fix documentation
a08a8fd decorators.py: Add yield_once doctest

Adrianzatreanu commented:
@rultor merge

Adrianzatreanu commented:
great work, @justuswilhelm

justuswilhelm commented:
😅 oh the wonders of Git

rultor commented:
> @rultor merge

@adrianzatreanu OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/8461-220827604)

rultor commented:
> @rultor merge

@adrianzatreanu Done! FYI, the full log is [here](http://www.rultor.com/t/8461-220827604) (took me 1min)
